core/store/src/trie/split_state.rs

```diff
@@ -35,21 +39,96 @@ impl Trie {
 }
 
 impl ShardTries {
+    pub fn apply_state_changes_to_split_states<'a>(
+        &self,
+        shard_uid: ShardUId,
+        state_root: StateRoot,
+        state_roots: &HashMap<ShardUId, StateRoot>,
+        changes: ConsolidatedStateChanges,
+        account_id_to_shard_id: &(dyn Fn(&AccountId) -> ShardUId + 'a),
+    ) -> Result<HashMap<ShardUId, TrieChanges>, StorageError> {
+        let trie = self.new_trie_update_view(shard_uid, state_root);
+        let mut trie_updates: HashMap<_, _> = self.get_trie_updates(state_roots);
+        let mut insert_receipts = Vec::new();
+        let mut delete_receipts = Vec::new();
+        for ConsolidatedStateChange { trie_key, value } in changes {
+            match trie_key {
+                TrieKey::DelayedReceiptIndices => {}
+                TrieKey::DelayedReceipt { index } => match value {
+                    Some(value) => {
+                        let receipt: Receipt = <_>::try_from_slice(&value).map_err(|_| {
+                            StorageError::StorageInconsistentState(format!(
+                                "invalid delayed receipt {:?}",
+                                value
+                            ))
+                        })?;
+                        insert_receipts.push((index, receipt));
+                    }
+                    None => {
+                        let receipt = get::<Receipt>(&trie, &TrieKey::DelayedReceipt { index })?
+                            .expect("deleted receipt does not exist");
+                        delete_receipts.push((index, receipt));
+                    }
+                },
+                _ => {
+                    // we can safely unwrap here because all other types of trie keys have account ids
+                    let account_id = trie_key.get_account().unwrap();
+                    let new_shard_uid = account_id_to_shard_id(account_id);
```

Comment:
```suggestion
                TrieKey::Account { account_id }
                | TrieKey::ContractCode { account_id }
                | TrieKey::AccessKey { account_id, .. }
                | TrieKey::ReceivedData { receiver_id, .. }
                | TrieKey::PostponedReceiptId { receiver_id, .. }
                | TrieKey::PendingDataCount { receiver_id, .. }
                | TrieKey::PostponedReceipt { receiver_id, .. }
                | TrieKey::ContractData { account_id, .. } => {
                    let new_shard_uid = account_id_to_shard_id(&account_id);
```

While the comment is true at this point in time, nothing prevents it from becoming false in the future, if we add a new `TrieKey`.